File: contracts/java/src/org/iota/wasp/wasmlib/host/Host.java
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package org.iota.wasp.wasmlib.host;

import de.mirkosertic.bytecoder.api.*;
import org.iota.wasp.wasmlib.keys.*;
import org.iota.wasp.wasmlib.mutable.*;

import java.nio.charset.*;

public class Host {
    public static final ScMutableMap root = new ScMutableMap(1);

    private static final byte[] TYPE_SIZES = {0, 33, 37, 0, 33, 32, 37, 32, 4, 8, 0, 34, 0};

    @Import(module = "WasmLib", name = "javaGetBytes")
    public static native int hostGetBytes(int objId, int keyId, int typeId, byte[] value, int size);

    @Import(module = "WasmLib", name = "javaGetKeyId")
    public static native int hostGetKeyId(byte[] key, int size);

    @Import(module = "WasmLib", name = "javaGetObjectId")
    public static native int hostGetObjectId(int objId, int keyId, int typeId);

    @Import(module = "WasmLib", name = "javaSetBytes")
    public static native void hostSetBytes(int objId, int keyId, int typeId, byte[] value, int size);

    public static void Clear(int objId) {
        SetBytes(objId, Key.Length.KeyId(), ScType.TYPE_INT64, new byte[8]);
    }

    public static boolean Exists(int objId, int keyId, int typeId) {
        // negative length (-1) means only test for existence
        // returned size -1 indicates keyId not found (or error)
        // this removes the need for a separate hostExists function
        return hostGetBytes(objId, keyId, typeId, null, -1) >= 0;
    }

    public static byte[] GetBytes(int objId, int keyId, int typeId) {
        // first query length of bytes array
        int size = hostGetBytes(objId, keyId, typeId, null, 0);
        if (size <= 0) {
            return new byte[TYPE_SIZES[typeId]];
        }

        // allocate a byte array in Wasm memory and
        // copy the actual data bytes to Wasm byte array
        byte[] bytes = new byte[size];
        //noinspection ResultOfMethodCallIgnored
        hostGetBytes(objId, keyId, typeId, bytes, size);
        return bytes;
    }

    public static int GetKeyIdFromBytes(byte[] bytes) {
        int size = bytes.length;
        // negative size indicates this was from bytes
        return hostGetKeyId(bytes, -size - 1);
    }

    public static int GetKeyIdFromString(String key) {
        byte[] bytes = key.getBytes(StandardCharsets.UTF_8);
        // non-negative size indicates this was from string
        return hostGetKeyId(bytes, bytes.length);
    }

    public static int GetLength(int objId) {
        byte[] bytes = GetBytes(objId, Key.Length.KeyId(), ScType.TYPE_INT64);
        return (bytes[0] & 0xff) | ((bytes[1] & 0xff) << 8) | ((bytes[2] & 0xff) << 16) | ((bytes[3] & 0xff) << 24);
    }

    public static int GetObjectId(int objId, int keyId, int typeId) {
        return hostGetObjectId(objId, keyId, typeId);
    }

    public static void Log(String text) {
        SetBytes(1, Key.Log.KeyId(), ScType.TYPE_STRING, text.getBytes(StandardCharsets.UTF_8));
    }

    public static void Panic(String text) {
        SetBytes(1, Key.Panic.KeyId(), ScType.TYPE_STRING, text.getBytes(StandardCharsets.UTF_8));
    }

    public static void SetBytes(int objId, int keyId, int typeId, byte[] value) {
        hostSetBytes(objId, keyId, typeId, value, value.length);
    }

    public static void Trace(String text) {
        SetBytes(1, Key.Log.KeyId(), ScType.TYPE_STRING, text.getBytes(StandardCharsets.UTF_8));
    }
}

File: packages/vm/wasmclient/host.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// +build wasm

package wasmclient

import "github.com/iotaledger/wasplib/packages/vm/wasmlib"

//go:wasm-module WasmLib
//export hostGetBytes
func hostGetBytes(objID, keyID, typeID int32, value *byte, size int32) int32

//go:wasm-module WasmLib
//export hostGetKeyID
func hostGetKeyID(key *byte, size int32) int32

//go:wasm-module WasmLib
//export hostGetObjectID
func hostGetObjectID(objID, keyID, typeID int32) int32

//go:wasm-module WasmLib
//export hostSetBytes
func hostSetBytes(objID, keyID, typeID int32, value *byte, size int32)

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type WasmVMHost struct{}

// implements wasmlib.ScHost interface
var _ wasmlib.ScHost = &WasmVMHost{}

func ConnectWasmHost() {
	wasmlib.ConnectHost(WasmVMHost{})
}

func (w WasmVMHost) CallFunc(objID, keyID int32, params []byte) []byte {
	args := (*byte)(nil)
	size := int32(len(params))
	if size != 0 {
		args = &params[0]
	}

	// pass params and query expected length of result
	size = hostGetBytes(objID, keyID, wasmlib.TYPE_CALL, args, size)

	// -1 means non-existent, so return default value for type
	if size <= 0 {
		return []byte(nil)
	}

	// allocate a sufficient length byte array in Wasm memory
	// and let the host copy the actual result into this Wasm byte array
	result := make([]byte, size)
	hostGetBytes(objID, keyID, wasmlib.TYPE_CALL+1, &result[0], size)
	return result
}

func (w WasmVMHost) Exists(objID, keyID, typeID int32) bool {
	// negative length (-1) means only test for existence
	// returned size -1 indicates keyID not found (or error)
	// this removes the need for a separate hostExists function
	return hostGetBytes(objID, keyID, typeID, nil, -1) >= 0
}

func (w WasmVMHost) GetBytes(objID, keyID, typeID int32) []byte {
	size := int32(wasmlib.TypeSizes[typeID])
	if size == 0 {
		// variable-sized type, first query expected length of bytes array
		// (pass zero-length buffer)
		size = hostGetBytes(objID, keyID, typeID, nil, 0)

		// -1 means non-existent, so return default value for type
		if size <= 0 {
			return []byte(nil)
		}
	}

	// allocate a sufficient length byte array in Wasm memory
	// and let the host copy the actual data bytes into this Wasm byte array
	result := make([]byte, size)
	hostGetBytes(objID, keyID, typeID, &result[0], size)
	return result
}

func (w WasmVMHost) GetKeyIDFromBytes(bytes []byte) int32 {
	size := int32(len(bytes))
	// &bytes[0] will panic on zero length slice, so use nil instead
	// negative size indicates this was from bytes
	if size == 0 {
		return hostGetKeyID(nil, -1)
	}
	return hostGetKeyID(&bytes[0], -size-1)
}

func (w WasmVMHost) GetKeyIDFromString(key string) int32 {
	bytes := []byte(key)
	size := int32(len(bytes))
	// &bytes[0] will panic on zero length slice, so use nil instead
	// non-negative size indicates this was from string
	if size == 0 {
		return hostGetKeyID(nil, 0)
	}
	return hostGetKeyID(&bytes[0], size)
}

func (w WasmVMHost) GetObjectID(objID, keyID, typeID int32) int32 {
	return hostGetObjectID(objID, keyID, typeID)
}

func (w WasmVMHost) SetBytes(objID, keyID, typeID int32, value []byte) {
	// &bytes[0] will panic on zero length slice, so use nil instead
	size := int32(len(value))
	if size == 0 {
		hostSetBytes(objID, keyID, typeID, nil, size)
		return
	}
	hostSetBytes(objID, keyID, typeID, &value[0], size)
}
